stellargraph/layer/graph_classification.py

```diff
@@ -0,0 +1,236 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from tensorflow.keras import activations, initializers, constraints, regularizers
+from tensorflow.keras.layers import (
+    Input,
+    Layer,
+    Lambda,
+    Dropout,
+    Reshape,
+    Masking,
+    GlobalAveragePooling1D,
+)
+
+from ..mapper import GraphGenerator
+from .cluster_gcn import ClusterGraphConvolution
+from ..core.experimental import experimental
+
+
+@experimental(reason="Missing unit tests")
+class GraphClassificationConvolution(ClusterGraphConvolution):
```

Comment:
I'm a bit confused by this class: its docs talk about being based on a GCN layer, but... its actually based on a Cluster-GCN one. In addition (and more importantly), it has a custom `call` function. Could you describe a bit about why it needs to differ to the base layer?

Reply:
The `ClusterGraphConvolution` layer is a GCN layer. It exists because of limitations/restrictions on our `GraphConvolution`, i.e., GCN, and `GraphAttention`, i.e, GAT, layers. The GAT and GCN layers require that graph size is fixed (see #1038) and also expect that batch size = 1 (and basically treat it as a nuisance that must be removed and added it to keep Keras happy). Further, one of the inputs indicates node indices, `out_indices`, that the last layer uses to `gather` the outputs just for the nodes of interest, i.e., those nodes that have targets (for training and evaluation) or the nodes we want predictions for; before `gather` was introduced our GCN and GAT models made predictions for all nodes in the graph regardless of the values passed to `generator.flow()` making the API inconsistent with methods such as GraphSAGE. 

When I implemented ClusterGCN, I wanted to use our existing GCN layer but the above issues that we did not have a solution for at the time forced me to create the new `ClusterGraphConvolution` layer which is essentially a more flexible GCN layer that does not support sparse inputs but can handle graphs of different sizes.

For graph classification, I wanted to re-use our GCN and GAT layers but the above problems once again made it difficult. The closest I had to a re-usable layer was `ClusterGraphConvolution` but it wasn't exactly right. The problem is that the input for graph classification does not include node indices, i.e., `out_indices`, since we need the output for all nodes for mean pooling. In graph classification `out_indices` has been replaced by `mask` to handle different size graphs. Also, even though Cluster-GCN is a mini-batch method, its definition of a batch is a bit different than in the graph case. Basically, the input for Cluster-GCN is still a single graph hence batch size = 1 and so the layer treats it like a nuisance to remove and add before and after calculation. The `GraphClassificationConvolution` layer receives data with batch size >= 1 and handles them correctly with a call to `keras.batch_dot()`.  The change to make the latter work was only necessary in the `call` method so the least I had to do to make it work is sub-class `ClusterGraphConvolution` and overwrite the `call` method. 

If/when we manage to fix our basic GCN and GAT layers, we can replace `GraphClassificationConvolution` with one of them and changes should be minimal. If #1038 is resolved, then we are going to be half way there; we would still have to deal with the batch dimension correctly; I think we might be able to fix it using `batch_dot` but we also have to check that `gather` works with a batch size dimension which is not clear from the spartan documentation at https://www.tensorflow.org/api_docs/python/tf/keras/backend/gather.

Reply:
> The problem is that the input for graph classification does not include node indices, i.e., out_indices, since we need the output for all nodes for mean pooling. In graph classification out_indices has been replaced by mask to handle different size graphs

Other than the gather difficulties, _could_ we use `out_indices` rather than `mask`, and pass all of the nodes as indices?

> I think we might be able to fix it using batch_dot but we also have to check that gather works with a batch size dimension which is not clear from the spartan documentation at https://www.tensorflow.org/api_docs/python/tf/keras/backend/gather.

That is rather poor, especially because, based [on the source](https://github.com/tensorflow/tensorflow/blob/988ba54e96aa1c0254c00823049cf46109d4335c/tensorflow/python/keras/backend.py#L1936-L1965), it just defers to `tf.gather`, which has much better docs: https://www.tensorflow.org/api_docs/python/tf/gather

Although, given how few arguments the Keras one has, we may have to call `tf.gather` directly, if `axis`'s default isn't what we need.

> If #1038 is resolved

Minor, but that issue is now closed, without us having actually fixed GAT/GCN, so if we're going to want to do that work, we'll need a another issue to track it.

Reply:
Initially, I did consider using `out_indices` instead of `masks` but with having to pad the data matrices it is not clear to me what to pad the `out_indices` with since we need `out_indices` to have `batch size x max nodes` size or we can't stack them together for the batch. Looking at the documentation for `tf.gather`  and if I understand correctly, repeating an index makes extra copies of it which is not what we want. 

We are going to open a new issue to tackle the problems with our GAT (and I think also GCN) layer.

Reply:
> We are going to open a new issue to tackle the problems with our GAT (and I think also GCN) layer.

Just so I'm not making an assumption about the responsibilities and the task doesn't get dropped accidentally, "we" = you?

Reply:
I had a chat with Kevin with regards to whether we should re-open #1038 or create a new issue. We agreed to open a new ticket and Kevin did so last week #1076.